src/test/scala/chiselTests/DPISpec.scala

```diff
@@ -0,0 +1,181 @@
+package chiselTests.simulator
+
+import chisel3._
+import chisel3.experimental.ExtModule
+import chisel3.simulator._
+import chisel3.util.circt.dpi._
+import circt.stage.ChiselStage
+import chisel3.util.{HasExtModuleInline, HasExtModulePath, HasExtModuleResource}
+import org.scalatest.funspec.AnyFunSpec
+import org.scalatest.matchers.must.Matchers
+import org.scalatest.matchers.should.Matchers.convertToAnyShouldWrapper
+import chisel3.util.HasBlackBoxInline
+import svsim._
+
+private object EmitDPIImplementation {
+  def apply() = {
+    val dpiImpl = s"""
+                     |#include <stdint.h>
+                     |#include <iostream>
+                     |
+                     |extern "C" void hello()
+                     |{
+                     |    std::cout << "hello from c++\\n";
+                     |}
+                     |
+                     |extern "C" void add(int lhs, int rhs, int* result)
+                     |{
+                     |    *result = lhs + rhs;
+                     |}
+  """.stripMargin
+
+    class DummyDPI extends BlackBox with HasBlackBoxInline {
+      val io = IO(new Bundle {})
+      setInline("dpi.cc", dpiImpl)
+      setInline(s"$desiredName.sv", s"module $desiredName(); endmodule")
+    }
+    val dummy = Module(new DummyDPI)
+
+  }
+}
+
+class DPIIntrinsicTest extends Module {
+  val io = IO(new Bundle {
+    val a = Input(UInt(32.W))
+    val b = Input(UInt(32.W))
+    val add_clocked_result = Output(UInt(32.W))
+    val add_unclocked_result = Output(UInt(32.W))
+  })
+
+  EmitDPIImplementation()
+
+  // Void function
+  RawClockedVoidFunctionCall("hello")(clock, true.B)
+
+  // Stateless function with result
+  val result_clocked =
+    RawClockedNonVoidFunctionCall("add", UInt(32.W), Some(Seq("lhs", "rhs")), Some("result"))(clock, true.B, io.a, io.b)
+  val result_unclocked =
+    RawUnclockedNonVoidFunctionCall("add", UInt(32.W), Some(Seq("lhs", "rhs")), Some("result"))(true.B, io.a, io.b)
+
+  io.add_clocked_result := result_clocked
+  io.add_unclocked_result := result_unclocked
+}
+
+object Hello extends DPIClockedVoidFunctionImport {
+  val functionName = "hello"
+  final def apply() = super.call()
+}
+
+object AddClocked extends DPINonVoidFunctionImport[UInt] {
+  override val functionName = "add"
+  val ret = UInt(32.W)
+  val clocked = true
```

Comment:
Should these also be `override`?

Reply:
They are abstract class members so override is optional but I'll add for the consistency, thanks!

Reply:
Stylistically I think we prefer to add `override` to call out that it's implementing a virtual method (but this is a relatively recent stylistic opinion so the codebase is pretty inconsistent). Maybe we can do this with a lint or rewrite rule...